src/ion_data/ion_eq.rs

```diff
@@ -22,39 +22,36 @@ use num_traits::Zero;
 ///    * `nan` and `nan` are Ion equivalent but not mathematically equivalent.
 ///    * `0.0e` and `-0.0e` are mathematically equivalent but not Ion equivalent.
 /// * Decimal `0.0` and `-0.0` are mathematically equivalent but not Ion equivalent.
+/// * Decimal `0.0` and `0.00` are mathematically equivalent but not Ion equivalent.
 /// * Timestamps representing the same point in time at different precisions or at different
 /// timezone offsets are not Ion equivalent.
 ///
 pub trait IonEq {
     fn ion_eq(&self, other: &Self) -> bool;
 }
 
-impl<T: IonEq> IonEq for &T {
-    fn ion_eq(&self, other: &Self) -> bool {
-        T::ion_eq(self, other)
-    }
-}
+// impl<T: IonEq> IonEq for &T {
+//     fn ion_eq(&self, other: &Self) -> bool {
+//         T::ion_eq(self, other)
+//     }
+// }
+//
+// impl<T: IonEq> IonEq for Vec<T> {
+//     fn ion_eq(&self, other: &Self) -> bool {
+//         self.as_slice().ion_eq(other.as_slice())
+//     }
+// }
 
-impl IonEq for bool {
+impl<R: Deref> IonEq for R
+where
+    <R as Deref>::Target: IonEq,
```

Comment:
Actually, there is a difference here, and it causes an error when I change it. Here's a minimum reproducible case:
```rust
#[test]
fn foo() {
    let element_vec: Vec<Element> = Element::read_all("foo 1 true").unwrap();
    bar(element_vec.into());
}

fn bar<T: IonOrd>(ion: IonData<T>) {
    // Do something
}
```
Fails with:
```
error[E0277]: the size for values of type `[element::Element]` cannot be known at compilation time
```
I'm assuming that this is because the trait impl is monomorphized differently depending on how the signature is formulated. (While rearranging a `where` clause shouldn't do something like this, in this case we have added a second type parameter _and_ we have changed how `Deref::Target` is specified.)

Do you have any insight or theory on the details of why this occurs?

Reply:
Aha... it needs `?Sized`:
```rust
impl<E, R> IonEq for R where
  E: IonEq + ?Sized,
  R: Deref<Target = E>,
```

Reply:
I ended up sticking with my original factoring because it minimizes the number of type parameters, though I cleaned it up by getting rid of some redundant parts.
```rust
impl<R: Deref> IonEq for R
where
    R::Target: IonEq,
```